src/algorithms/semihdp_sampler.cpp

```diff
@@ -0,0 +1,713 @@
+#include "semihdp_sampler.hpp"
+
+#include <proto/cpp/marginal_state.pb.h>
+
+#include <algorithm>
+#include <src/utils/eigen_utils.hpp>
+
+#include "../utils/distributions.hpp"
+
+SemiHdpSampler::SemiHdpSampler(const std::vector<MatrixXd>& data,
+                               std::shared_ptr<BaseHierarchy> hier,
+                               bayesmix::SemiHdpParams params)
+    : data(data), params(params) {
+  ngroups = data.size();
+  n_by_group.resize(ngroups);
+  for (int i = 0; i < ngroups; i++) n_by_group[i] = data[i].size();
+  G0_master_hierarchy = hier->clone();
+  G00_master_hierarchy = hier->clone();
+  alpha = params.alpha();
+  gamma = params.gamma();
+  a_w = params.w_prior().shape1();
+  b_w = params.w_prior().shape2();
+}
+
+void SemiHdpSampler::initialize() {
+  auto& rng = bayesmix::Rng::Instance().get();
+  omega = VectorXd::Ones(ngroups).array();
+
+  int INIT_N_CLUS = 5;
+  VectorXd probas = VectorXd::Ones(INIT_N_CLUS);
+  probas /= (1.0 * INIT_N_CLUS);
+  c.resize(ngroups);
+  is_used_c.resize(ngroups);
+  s.resize(ngroups);
+  v.resize(ngroups);
+  t.resize(ngroups);
+  theta_star.resize(ngroups);
+  theta_tilde.resize(ngroups);
+  n_by_theta_star.resize(ngroups);
+  theta_star_pseudo.resize(ngroups);
+  n_by_theta_star_pseudo.resize(ngroups);
+
+  for (int l = 0; l < INIT_N_CLUS; l++) {
+    std::shared_ptr<BaseHierarchy> hierarchy = G00_master_hierarchy->clone();
+    hierarchy->draw();
+    taus.push_back(hierarchy);
+  }
+
+  for (int i = 0; i < ngroups; i++) {
+    c[i] = i;
+    is_used_c[i] = true;
+    s[i].resize(n_by_group[i]);
+
+    for (int j = 0; j < INIT_N_CLUS; j++) s[i][j] = j;
+    for (int j = INIT_N_CLUS; j < n_by_group[i]; j++) {
+      s[i][j] = bayesmix::categorical_rng(probas, rng);
+    }
+
+    v[i].resize(2 * INIT_N_CLUS);
+    t[i].resize(2 * INIT_N_CLUS);
+
+    for (int l = 0; l < INIT_N_CLUS; l++) {
+      std::shared_ptr<BaseHierarchy> hierarchy = G0_master_hierarchy->clone();
+      hierarchy->draw();
+      theta_tilde[i].push_back(hierarchy);
+      theta_star[i].push_back(hierarchy);
+      v[i][l] = l;
+      t[i][l] = -1;
+    }
+
+    for (int l = 0; l < INIT_N_CLUS; l++) {
+      theta_star[i].push_back(taus[l]);
+      v[i][INIT_N_CLUS + l] = -1;
+      t[i][INIT_N_CLUS + l] = l;
+    }
+  }
+  m = std::vector<int>(taus.size(), 0);
+
+  for (int i = 0; i < ngroups; i++) {
+    n_by_theta_star[i] = std::vector<int>(theta_star[i].size(), 0);
+    for (int j = 0; j < n_by_group[i]; j++) {
+      n_by_theta_star[i][s[i][j]] += 1;
+    }
+  }
+}
+
+void SemiHdpSampler::update_unique_vals() {
+  // Loop through the theta stars, if its equal to some theta_tilde, we
+  // update it on the fly. Otherwise we store the the data in a vector
+  // and then update all the taus
+
+  std::vector<MatrixXd> data_by_tau(taus.size());
+  for (int h = 0; h < taus.size(); h++) {
+    data_by_tau[h] = Eigen::MatrixXd(0, 1);
```

Comment:
What is a Eigen::MatrixXd(0, 1)? What does this line do?

Reply:
I need this line because by default Eigen::MatrixXd is un-initialized, so that if you have
`auto a = Eigen::MatrixXd` and then ask `a.rows()` you could get values different from zero.
I need to set the rows to zero because then I want to append stuff to this (empty) matrix using `bayesmix::append_by_row`